saleor/graphql/product/filters.py

```diff
@@ -331,18 +331,13 @@ def filter_products_by_collections(qs, collection_pks):
 
 
 def filter_products_by_stock_availability(qs, stock_availability, channel_slug):
-    allocations = (
-        Allocation.objects.values("stock_id")
-        .filter(quantity_allocated__gt=0, stock_id=OuterRef("pk"))
-        .values_list(Sum("quantity_allocated"))
-    )
-    allocated_subquery = Subquery(queryset=allocations, output_field=IntegerField())
-
     stocks = (
         Stock.objects.for_channel_and_country(channel_slug)
-        .filter(quantity__gt=Coalesce(allocated_subquery, 0))
+        .annotate_available_quantity_including_reservations()
```

Comment:
@zedzior Did you test it with any big database? In my opinion, it's break optimization. When we use annotation then a new virtual table is made by Postgress.  Annotations shouldn't be used for filters. 
To confirm it you could generate explain.